scipy/ndimage/_filters.py

```diff
@@ -1486,8 +1487,28 @@ def _rank_filter(input, rank, size=None, footprint=None, output=None,
                 "A sequence of modes is not supported by non-separable rank "
                 "filters")
         mode = _ni_support._extend_mode_to_code(mode)
-        _nd_image.rank_filter(input, rank, footprint, output, mode, cval,
-                              origins)
+        if input.ndim == 1:
+            rank = int(rank)
+            origin = int(origin)
+            # legacy mode handling
+            if mode == 6:
+                mode = 4
+            if mode == 5:
+                mode = 1
+            casting_cond = input.dtype.name not in ['int64', 'float64', 'float32']
+            if casting_cond:
+                x = input.astype('int64')
+                x_out = np.empty_like(x)
```

Comment:
It looks to me like this should be using `np.result_type` to determine what the correct output type is. E.g., for other floating-point dtypes, `float32` or `float64` will be expected, not `int64`.